examples/vae_simple.py

```diff
@@ -0,0 +1,322 @@
+    ret.paste(orig, (0,0))
+    new = Image.fromarray(get_image_array(Xpr, i, shp, channels), mode=mode)
+    ret.paste(new, (0, orig.size[1]))
+    return ret
+
+# ############################# Batch iterator ###############################
+
+def iterate_minibatches(inputs, targets, batchsize, shuffle=False):
+    assert len(inputs) == len(targets)
+    if shuffle:
+        indices = np.arange(len(inputs))
+        np.random.shuffle(indices)
+    for start_idx in range(0, len(inputs) - batchsize + 1, batchsize):
+        if shuffle:
+            excerpt = indices[start_idx:start_idx + batchsize]
+        else:
+            excerpt = slice(start_idx, start_idx + batchsize)
+        yield inputs[excerpt], targets[excerpt]
+
+
+# ##################### Custom layer for middle of VCAE ######################
+# This layer takes the mu and sigma (both DenseLayers) and combines them with
+# a random vector epsilon to sample values for a multivariate Gaussian
+
+class GaussianSampleLayer(MergeLayer):
+    def __init__(self, mu, logsigma, rng=None, **kwargs):
+        self.rng = rng if rng else RandomStreams(nn.random.get_rng().randint(1,2147462579))
+        super(GaussianSampleLayer, self).__init__([mu, logsigma], **kwargs)
+
+    def get_output_shape_for(self, input_shapes):
+        return input_shapes[0]
+
+    def get_output_for(self, inputs, **kwargs):
+        mu, logsigma = inputs
+        shape=(self.input_shapes[0][0] or inputs[0].shape[0],
+                self.input_shapes[0][1] or inputs[0].shape[1])
+        return mu + T.exp(logsigma) * self.rng.normal(shape)
+
+
+# ############################## Build Model #################################
+# encoder has 1 hidden layer, where we get mu and sigma for Z given an inp X
+# continuous decoder has 1 hidden layer, where we get mu and sigma for X given code Z
+# binary decoder has 1 hidden layer, where we calculate p(X=1)
+# once we have (mu, sigma) for Z, we sample L times
+# Then L separate outputs are constructed and the final layer averages them
+
+def build_vae(inputvar, L=1, binary=False, imgshape=(28,28), channels=1, z_dim=2, n_hid=256):
+    x_dim = imgshape[0] * imgshape[1] * channels
+    l_input = nn.layers.InputLayer(shape=(None,channels,imgshape[0], imgshape[1]),
+            input_var=inputvar, name='input')
+    l_enc_hid = nn.layers.DenseLayer(l_input, num_units=n_hid,
+            nonlinearity=T.nnet.softplus, name='enc_hid')
+    l_enc_mu = nn.layers.DenseLayer(l_enc_hid, num_units=z_dim,
+            nonlinearity = None, name='enc_mu')
+    l_enc_logsigma = nn.layers.DenseLayer(l_enc_hid, num_units=z_dim,
+            nonlinearity = None, name='enc_logsigma')
+    l_Z_list = []
+    l_dec_mu_list = []
+    l_dec_logsigma_list = []
+    l_output_list = []
+    # tie the weights of all L versions so they are the "same" layer
+    W_dec_hid = None
+    b_dec_hid = None
+    W_dec_mu = None
+    b_dec_mu = None
+    W_dec_ls = None
+    b_dec_ls = None
+    for i in xrange(L):
```

Comment:
It would be more elegant to use a `RepeatLayer` to replicate your data points L times, then have a single stack of what you have below, and in the end use a `ReshapeLayer` and a `FeaturePoolLayer` to average over the L samples per data point. However, there is no `RepeatLayer` in Lasagne yet (https://github.com/Lasagne/Lasagne/issues/362), and it would require refactoring quite a lot of code. Maybe that's something for a future PR.